two small changes to Path parsing (fixes #257)

mossprescott opened the pull request:
Don't strip trailing `/`s in Path. Formerly, `Root / "foo"` would match
both `/foo` and `/foo/`. Now, you can/must do `Root / "foo" / ""` to
match the latter.

*Note: this change seems likely to break user code, but did not break any existing test.*

URL-decode path segments, so that it's possible to encode characters
like `/` and space.

bryce-anderson commented on dsl/src/main/scala/org/http4s/dsl/Path.scala:
```diff
@@ -31,11 +32,8 @@ object Path {
-        prefix
-      else
-        prefix / str.substring(slash + 1)
+      val prefix = Path(UrlCodingUtils.urlDecode(str.substring(0, slash)))
```
I think I see a bug that will cause problems with ```%2F``` in the path elements. I see you have a test for this below in the form of ```"/foo%20bar/1%2F2"``` that passes, but what about ```"/foo%2Fbar/blah"```? I suspect that will end up structured as ```Root / "foo" / "bar" / "blah"``` as opposed to the intended ```Root / "foo/bar" / "blah"```. If so, the culprit is urldecoding the argument here.
If not, my apologies :smile:

mossprescott replied:
You're right, there is a bug there. I'll add a test and fix it. Thanks for
the review!

bryce-anderson commented:
This seems simple enough. Once my comment is addressed I'm fine with merging this. Its too bad that it will change the default behavior, but such is life: I like this behavior. We should probably look into making a middleware that can swallow the trailing ```/``` for those that want the old behavior, but thats another issue.

@mossprescott, thanks for the PR.

mossprescott commented:
Fixed and added that pattern to the test. While I was in there testing it, I realized that `Path.toString` was producing confusingly unencoded results, so I fixed that as well (and added a very simple test). Hope I didn't introduce another bug.

bryce-anderson commented:
Cool, that looks good to me. I'll give this until tomorrow to merge this to let @rossabaker weigh in if he wants. Thanks again for the PR. 
As a small bit of self promotion, you may be interested in the http4s/rho project. Maybe not. :smile:
